be/src/exec/pipeline/pipeline_driver_poller.cpp

```diff
@@ -48,20 +48,24 @@ void PipelineDriverPoller::run_internal() {
             if (driver->pending_finish() && !driver->source_operator()->pending_finish()) {
                 // driver->pending_finish() return true means that when a driver's sink operator is finished,
                 // but its source operator still has pending io task that executed in io threads and has
-                // reference to object outside(such as desc_tbl) owned by FragmentContext. So an driver in
+                // reference to object outside(such as desc_tbl) owned by FragmentContext. So a driver in
                 // PENDING_FINISH state should wait for pending io task's completion, then turn into FINISH state,
                 // otherwise, pending tasks shall reference to destructed objects in FragmentContext since
                 // FragmentContext is unregistered prematurely.
-                driver->set_driver_state(DriverState::FINISH);
+                driver->set_driver_state(driver->fragment_ctx()->is_canceled() ? DriverState::CANCELED
+                                                                               : DriverState::FINISH);
                 _dispatch_queue->put_back(*driver_it);
                 local_blocked_drivers.erase(driver_it++);
             } else if (driver->is_finished()) {
                 local_blocked_drivers.erase(driver_it++);
-            } else if (driver->query_ctx()->is_expired()) {
-                // there are not any drivers belonging to a query context can make progress for a expiration period
+            } else if (!driver->pending_finish() && driver->query_ctx()->is_expired()) {
```

Comment:
Is it possible that driver's state being changed between the two conditions?
1. 「driver->pending_finish()」 turns out false
2. driver->_state becomes pending_finish
3. 「driver->query_ctx()->is_expired()」 turns out true

How to deal with it?